Sources/CodeGen/CXX/CodeFormatter.swift

```diff
@@ -0,0 +1,88 @@
+/// Class responsible for the formatiing of the CXX code, as it is generated.
+///
+/// The callers will tell this class what tokens they want to write to the output CXX code,
+/// and this will ensure that proper formatting will be added.
+///
+/// This class will have ownership of the code that is being written.
+struct CodeFormatter: TextOutputStream {
+
+  /// The amount of indentation we apply to the code.
+  private var indentSize: Int
+
+  /// The code that this object is writing.
+  var code: String = ""
+
+  /// The current level of indentation.
+  private var curIndent: Int = 0
+  /// True if the next text written should be indented.
+  private var shouldIntent: Bool = false
+
+  /// Initializes this object with the right formatting parameters.
+  init(indentSize: Int = 2) {
+    self.indentSize = indentSize
+  }
+
+  /// Adds new text to the output code.
+  mutating func write(_ text: String) {
+    checkAddIndent()
+    code.write(text)
+  }
+
+  /// Adds new text to the output code, with a new line at the end
+  mutating func writeLn(_ text: String) {
+    checkAddIndent()
+    code.write(text)
+    addNewLine()
+  }
+
+  /// Adds a space to the code that is currently written.
+  mutating func writeSpace() {
+    checkAddIndent()
+    code.write(" ")
+  }
+
+  /// Write a new line to the current code.
+  /// If the next line contains visible characters, they would be indented.
+  mutating func addNewLine() {
```

Comment:
```suggestion
  mutating func writeNewline() {
```

1. We don't capitalize "line" in "newline" in other parts of the codebase.
2. The function should be named `writeNewline` for consistency
3. As mentioned above, I would prefer `write("\n")` for an economy of concepts